apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.beam.sdk.options.Default;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.lang.reflect.Type;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class CtrPrediction {
+
+    static class FeatureMap extends DoFn<String, String> {
+
+        private final Map<String, Map<String, String>> featureMap = readFeatureMap();
+        private final int numFeatures = featureMap.size();
+
+        @ProcessElement
+        public void processElement(@Element String element, OutputReceiver<String> receiver) {
+            String[] fields = element.trim().split("\t");
+            List<String> features = new ArrayList<>();
+            // first column uuid, second column label (ignored)
+            if (fields.length == numFeatures + 2) {
+                features.add(fields[0]);
+                for (int i = 1; i < numFeatures + 1; i++) {
+                    features.add(featureMap.get(String.valueOf(i)).getOrDefault(fields[i + 1].trim(), "-1"));
+                }
+                receiver.output(String.join("\t", features));
+            }
+        }
+    }
+
+
+    public static void main(String[] args) throws MalformedModelException, ModelNotFoundException, IOException {
+        // creates pipeline from arguments
+        CtrOptions options =
+                PipelineOptionsFactory.fromArgs(args).withValidation().as(CtrOptions.class);
+        Pipeline p = Pipeline.create(options);
+        PCollection<String> records = p.apply("ReadData", TextIO.read().from(options.getInputFile()));
+        // add unique id to each record
+        PCollection<String> addIds = records.apply("AddUUID", ParDo.of(new DoFn<String, String>() {
+            @ProcessElement
+            public void processElement(ProcessContext c) {
+                String uniqueID = UUID.randomUUID().toString();
+                c.output(uniqueID + "\t" + c.element());
+            }
+        }));
+
+        // convert categorical features to integer according to feature map used in training
+        PCollection<String> preprocess = addIds.apply("Preprocess", ParDo.of(new FeatureMap()));
+
+        // run inference using Deep Java Library
+        PCollection<String> ctr = preprocess.apply("Inference", ParDo.of(new DoFn<String, String>() {
+            @ProcessElement
+            public void processElement(ProcessContext c) throws TranslateException, MalformedModelException, ModelNotFoundException, IOException {
+                Criteria<String, String> criteria = Criteria.builder()
```

Comment:
We are loading the model on each process, we should move this to class member

Reply:
Our predictor is not serializable, I will see if I can update the DJL side

Reply:
mvn, fixed by making it static